content/advent-2016/goroutree.md

```diff
@@ -0,0 +1,285 @@
+
+I show a bit of code, but most of my descriptions are in prose because code would make the post get
+quite long. if you want to follow along in the code, then open the link at the bottom to the GitHub
+repository that hold the package.
+
+# Example
+
+Let's take a look at a usage example right off the bat:
+
+```go
+import "github.com/ScottMansfield/goroutree"
+
+func main() {
+    g := goroutree.New()
+
+    boolres := make(chan bool)
+    g.Insert(boolres, 5)
+    <-boolres // will return true
+
+    g.Contains(boolres, 5)
+    <-boolres // will return true
+
+    g.Contains(boolres, 4)
+    <-boolres // will return false
+
+    g.Delete(boolres, 4)
+    <-boolres // will return false
+
+    g.Delete(boolres, 5)
+    <-boolres // will return true
+}
+```
+
+# Goroutree
+
+The `Goroutree` type is very simple:
+
+```go
+type Goroutree struct {
+    cmdchan chan cmd
+}
+```
+
+It only has a command channel to send commands to and nothing else. The `New()`
+function only needs to start up the manager goroutine and return a pointer to
+the `Goroutree`.
+
+```go
+func New() *Goroutree {
+    cmdchan := make(chan cmd)
+    go manager(cmdchan)
+
+    return &Goroutree{cmdchan}
+}
+```
+
+# Manager Goroutine
+
+There's actually a hidden "super root" node that takes the initial requests and then forwards them
+as needed to the root node. This is to support things like having an empty tree or deleting the root
+node. For example, when checking if the tree contains a number, if the root node is null the answer
+is always false. For deleting the root node, someone has to be notified of the new root's handle
+(read: input channel). THe manager goroutine is there to play that role.
+
+Introducing the manager also made the synchronous interface be completely separated from the tree
+instead of sometimes requiring a new goroutine to be spawned just to respond to the request.
+
+# Node Goroutine
+
+Each node is a separate goroutine that owns a particular value. It has an input channel that is a
+stream of commands which the node will react to one at a time.
+
+Nodes have knowledge of their own value (an int), and several channels:
+
+* input
+* parent
+* left child
+* right child
+
+The input and parent channels are guaranteed to be not nil, but the left and right are not. Each
+node is independently and concurrently receiving and serving requests coming in. There's some
+situations where a child will message a parent, but mostly the commands go downward.
+
+# Operations
+
+There's three main operations on the set: `Insert`, `Contains`, and `Delete`. There is a fourth,
+`Print`, that is mainly used for debugging.
+
+The operations all follow the basic binary search tree patterns. There's no facny rotations that
+happen in the middle, but there are a couple somewhat complicated interactions during deletion of a
+node. There's some extra commands that are only used internally that I will show as well.
```

Comment:
nit: I think `There's some extra commands` should be `There are some extra commands`.